### PR TITLE
Adds experimental Liger support to SFT script

### PR DESCRIPTION
**Pull request by edbeeching**
# What does this PR do?

Adds experimental liger-kernel support to TRL's SFT script:
##Usage (tested on 8 A100s with deepspeed)
### Without liger (OOMs)
```
accelerate launch --config_file=examples/accelerate_configs/deepspeed_zero3.yaml examples/scripts/sft_liger.py --model_name_or_path meta-llama/Meta-Llama-3.1-8B --dataset_text_field="text" --report_to="wandb" --learning_rate=2.0e-5 --per_device_train_batch_size=64 --max_seq_length=512 --output_dir meta-llama/Meta-Llama-3.1-8B-no-liger --gradient_checkpointing --attn_implementation flash_attention_2 --gradient_checkpointing_kwargs='{"use_reentrant":"True"}' --bf16=True --logging_steps=5 --use_liger=False
```

### With Liger
```
accelerate launch --config_file=examples/accelerate_configs/deepspeed_zero3.yaml examples/scripts/sft_liger.py --model_name_or_path meta-llama/Meta-Llama-3.1-8B --dataset_text_field="text" --report_to="wandb" --learning_rate=2.0e-5 --per_device_train_batch_size=64 --max_seq_length=512 --output_dir meta-llama/Meta-Llama-3.1-8B-liger --gradient_checkpointing --attn_implementation flash_attention_2 --gradient_checkpointing_kwargs='{"use_reentrant":"True"}' --bf16=True --logging_steps=5 --use_liger=True
```

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -82,6 +82,7 @@
         "llm-blender>=0.0.2",
     ],
     "peft": ["peft>=0.8.0"],
+    "liger": ["liger-kernel-nightly==0.1.1.dev20240828063745"],
```
Comment (edbeeching) [inline setup.py:85]: I am not a big fan of pinning this dev version so we should update to the real package ASAP·

File: trl/commands/cli_utils.py [hunk 2]
```diff
@@ -89,6 +89,7 @@ class SFTScriptArguments:
         default=False,
         metadata={"help": "Whether to apply `use_reentrant` for gradient_checkpointing"},
     )
+    use_liger: bool = field(default=False, metadata={"help": "Monkey patch the model with Liger kernels"})
```
Comment (lewtun) [inline trl/commands/cli_utils.py:92]:
```suggestion
    use_liger: bool = field(default=False, metadata={"help": "Monkey patch the model with Liger kernels to increase throughput and reduce memory usage."})
```

**Review by lewtun — APPROVED** [review]
Very clean addition! LGTM with a small nit :)

**Comment by lewtun** [comment]
@kashif correctly pointed out that this flag should live in the `SFTConfig` instead of the `SFTTrainer` init - would you mind doing that?

**Review by kashif — APPROVED** [review]

File: examples/scripts/sft.py [hunk 3]
```diff
@@ -143,6 +143,7 @@
             tokenizer=tokenizer,
             peft_config=get_peft_config(model_config),
             callbacks=[RichProgressCallback] if TRL_USE_RICH else None,
+            use_liger=args.use_liger,
```
Comment (lewtun) [inline examples/scripts/sft.py:146]: This should be removed right?

File: trl/trainer/sft_trainer.py [hunk 4]
```diff
@@ -140,6 +143,7 @@ def __init__(
         model_init_kwargs: Optional[Dict] = None,
         dataset_kwargs: Optional[Dict] = None,
         eval_packing: Optional[bool] = None,
+        use_liger: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (lewtun) [reply trl/trainer/sft_trainer.py:146]: Yes I think it should be removed now that it is in the config right?

File: setup.py (hunk 1 above)
Comment (ByronHsu) [reply setup.py:85]: we will cut a release in few hours

**Comment by ByronHsu** [comment]
Will we add some documentation on sft website too? that is where people learn how to use sft

**Comment by shimizust** [comment]
@edbeeching We previously added the `use_liger_kernel` flag to transformers TrainingArgs ([PR](https://github.com/huggingface/transformers/pull/32860/files#diff-bfceaff300c851b8e24fc50dc6638482abaec8f7d2a718e877c3828c166bcf79L1495)) which SFTConfig extends, but that hasn't been released yet. I'm thinking this should reuse that argument?

@ArtherZucker do you think the latest transformers version can be released so this flag can be used instead?

**Pull request merged**